### PR TITLE
Add `inngest/function.failed` internal event for fn failures

### PR DESCRIPTION
**Pull request by jpwilliams**
## Summary

Adds the sending of an internal `inngest/function.failed` event when a function fails for the final time.

Unsure about the required use of `EventStream` here; it could be better.

## Todo

- [x] Create `WithFailureHandler()` for when instantiating an executor - takes a function to call when a run has failed for the final time
- [x] Ensure executor has no knowledge of required pubsub/eventstream/etc
- [ ] Tests

## Related

- inngest/inngest-js#133

### CODE REVIEW

File: pkg/execution/executor/executor.go [hunk 1]
```diff
@@ -146,6 +146,13 @@ func WithConfig(c config.Execution) ExecutorOpt {
 	}
 }
 
+func WithFailureHandler(f *func(context.Context, state.Identifier, state.State, state.DriverResponse) error) ExecutorOpt {
```
Comment (tonyhb) [inline pkg/execution/executor/executor.go:149]: We could type this handler for readability/ease of use:  you can then use this type here and in the struct definition.


```
type FailureHandler func(context.Context, state.Identifier, state.State, state.DriverResponse) error

func WithFailureHandler(f FailureHandler) ExecutorOpt {
```

File: pkg/execution/executor/executor.go [hunk 2]
```diff
@@ -173,6 +180,7 @@ type executor struct {
 	sm             state.Manager
 	al             coredata.ExecutionActionLoader
 	runtimeDrivers map[string]driver.Driver
+	failureHandler *func(context.Context, state.Identifier, state.State, state.DriverResponse) error
```
Comment (tonyhb) [inline pkg/execution/executor/executor.go:183]: This doesn't need to be a pointer, esp. when we convert this to a type :)

File: pkg/execution/executor/executor.go [hunk 3]
```diff
@@ -306,6 +314,12 @@ func (e *executor) run(ctx context.Context, w inngest.Workflow, id state.Identif
 		return nil, idx, err
 	}
 	if response.Err != nil {
+		if response.Final() && e.failureHandler != nil {
+			if err := (*e.failureHandler)(ctx, id, s, *response); err != nil {
```
Comment (tonyhb) [inline pkg/execution/executor/executor.go:318]: When this isn't a pointer you can literally just `if err := e.failureHandler(ctx, id, s, *response)` :)

File: pkg/execution/state/driver_response.go [hunk 4]
```diff
@@ -236,3 +237,46 @@ func (r DriverResponse) Error() string {
...
+	}
+
+	var output UserError
+
+	if err := json.Unmarshal([]byte(s), &output); err != nil {
+		return nil
+	}
+
+	return &output
+}
```
[the remark this answers is not in the document]
Comment (tonyhb) [reply pkg/execution/state/driver_response.go:282]: I hate the way we do errors so this isn't on you.  Let's standardize next quarter.  We could pair on a unit test on Monday!

**Review by tonyhb — CHANGES_REQUESTED** [review]

File: pkg/execution/executor/executor.go (hunk 1 above)
Comment (jpwilliams) [reply pkg/execution/executor/executor.go:149]: Lovely!

File: pkg/execution/executor/executor.go (hunk 2 above)
Comment (jpwilliams) [reply pkg/execution/executor/executor.go:183]: @tonyhb Figured that this `WithX` pattern is for adding items that aren't required; wouldn't making this a value mean that it's always needed?

Just a meta question; I'm aware we always want it in this particular case!

**Comment by jpwilliams** [comment]
Looking more reasonable, @tonyhb?

**Review by tonyhb — APPROVED** [review]

**Comment by tonyhb** [comment]
Absolutely, this looks great!

**Pull request merged**